src/metrics.jl

```diff
@@ -180,70 +209,97 @@ end
             @simd for I in eachindex(a, b)
                 ai = a[I]
                 bi = b[I]
-                s = eval_reduce(d, s, eval_op(d, ai, bi))
+                s = eval_reduce(d, s, eval_op(d, ai, bi, I))
```

Comment:
Actually, seeing the code I understand better what you're doing here. How about defining a `parameters(d)` method instead which would return the parameter vector? A fallback method would return `nothing` for most distances, and in that case you'd call the two-argument `eval_op`. That way indexing is clearly separated from actually computing the distance, most distances don't need any changes and you don't need to introduce the `ParametrizedMetrics` union.

Reply:
Sounds good, I'll give it a try.